sklearn/tree/_utils.pyx

```diff
@@ -22,22 +22,20 @@ from ..utils._random cimport our_rand_r
 # Helper functions
 # =============================================================================
 
-cdef realloc_ptr safe_realloc(realloc_ptr* p, size_t nelems) except * nogil:
+cdef int safe_realloc(realloc_ptr* p, size_t nelems) except -1 nogil:
```

Comment:
Since we never use the result of this function (it's done inplace), the simplest solution to propagate exception without acquiring the gil every time is to make it return an int.